internal/storage/bucket_handle.go

```diff
@@ -96,3 +96,40 @@ func (b *bucketHandle) StatObject(ctx context.Context, req *gcs.StatObjectReques
 
 	return
 }
+
+func (bh *bucketHandle) CreateObject(ctx context.Context, req *gcs.CreateObjectRequest) (o *gcs.Object, err error) {
+	obj := bh.bucket.Object(req.Name)
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		if req.MetaGenerationPrecondition != nil && *req.MetaGenerationPrecondition != 0 {
+			obj = obj.If(storage.Conditions{GenerationMatch: *req.GenerationPrecondition, MetagenerationMatch: *req.MetaGenerationPrecondition})
+		} else {
+			obj = obj.If(storage.Conditions{GenerationMatch: *req.GenerationPrecondition})
+		}
+	}
+
+	// Creating a NewWriter with requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	wc.ChunkSize = 0 // This enables one shot upload.
+	wc = storageutil.SetAttrsInWriter(wc, req)
+
+	// Copy the contents to the writer.
+	if _, err = io.Copy(wc, req.Contents); err != nil {
+		err = fmt.Errorf("error in io.Copy: %w", err)
+		return
+	}
+
+	// We can't use defer to close the writer, because we need to close the
+	// writer successfully before calling Attrs() method of writer.
+	if err = wc.Close(); err != nil {
```

Comment:
just curious, when exactly does write to gcs happens, when do you wc.close or io.copy?

Reply:
io.copy writes the data to gcs (internally it open writer method of httpStorageClient, which creates pipe to write it to the gcs), and close method completes the write operation and flush any remaining buffered data in writer-pipe.